plugin/trino-base-jdbc/src/test/java/io/trino/plugin/jdbc/BaseJdbcCastPushdownTest.java

```diff
@@ -58,37 +56,43 @@ public void testProjectionPushdownWithCast()
     public void testJoinPushdownWithCast()
     {
         for (CastTestCase testCase : supportedCastTypePushdown()) {
-            assertThat(query("SELECT l.id FROM %s l JOIN %s r ON CAST(l.%s AS %s) = r.%s".formatted(leftTable(), rightTable(), testCase.sourceColumn(), testCase.castType(), testCase.targetColumn().orElseThrow())))
+            assertThat(query("SELECT l.id FROM %s l JOIN %s r ON CAST(l.%s AS %s) = r.%s".formatted(leftTable(), rightTable(), testCase.sourceColumn(), testCase.castType(), testCase.targetColumn())))
                     .isFullyPushedDown();
         }
 
         for (CastTestCase testCase : unsupportedCastTypePushdown()) {
-            assertThat(query("SELECT l.id FROM %s l JOIN %s r ON CAST(l.%s AS %s) = r.%s".formatted(leftTable(), rightTable(), testCase.sourceColumn(), testCase.castType(), testCase.targetColumn().orElseThrow())))
+            assertThat(query("SELECT l.id FROM %s l JOIN %s r ON CAST(l.%s AS %s) = r.%s".formatted(leftTable(), rightTable(), testCase.sourceColumn(), testCase.castType(), testCase.targetColumn())))
                     .joinIsNotFullyPushedDown();
         }
     }
 
     @Test
     public void testCastFails()
     {
-        for (CastTestCase testCase : failCast()) {
-            assertThatThrownBy(() -> getQueryRunner().execute("SELECT CAST(%s AS %s) FROM %s".formatted(testCase.sourceColumn(), testCase.castType(), leftTable())))
-                    .hasMessageMatching("(.*)Cannot cast (.*) to (.*)");
+        for (FailCastTestCase testCase : failCast()) {
+            assertThat(query("SELECT CAST(%s AS %s) FROM %s".formatted(testCase.sourceColumn(), testCase.castType(), leftTable())))
+                    .failure()
+                    .hasMessageContaining(testCase.message());
```

Comment:
This looks good! 
I wonder if we should complicate this a bit by adding second optional message so that if 2 messages are present (implying Trino and target database errors will be different) we could assert both with&without the pushdown, so something like this:

```java            
            Session withoutPushdown = Session.builder(getSession())
                    .setSystemProperty("allow_pushdown_into_connectors", "false")
                    .build();
            if (testCase.customPushdownMessage().isPresent()) {
                assertThat(query("SELECT CAST(%s AS %s) FROM %s".formatted(testCase.sourceColumn(), testCase.castType(), leftTable())))
                        .failure()
                        .hasMessageMatching(testCase.customPushdownMessage().get());
                assertThat(query(withoutPushdown, "SELECT CAST(%s AS %s) FROM %s".formatted(testCase.sourceColumn(), testCase.castType(), leftTable())))
                        .failure()
                        .hasMessageMatching(testCase.message());
            }
            else {
                assertThat(query("SELECT CAST(%s AS %s) FROM %s".formatted(testCase.sourceColumn(), testCase.castType(), leftTable())))
                        .failure()
                        .hasMessageMatching(testCase.message());
            }
```

While extending record constructors to be:

```java
        public FailCastTestCase(String sourceColumn, String castType)
        {
            this(sourceColumn, castType, "Cannot cast (.*) to (.*)", Optional.empty());
        }
        
        public FailCastTestCase(String sourceColumn, String castType,  @Language("RegExp") String customPushdownMessage)
        {
            this(sourceColumn, castType, "Cannot cast (.*) to (.*)", Optional.of(customPushdownMessage));
        }
```

WDYT?

Reply:
Thanks @SemionPar for the review.

I think this will help in redshift as well for cases like. However we might need to introduce WHERE clause also later.

```
assertThat(query(sessionWithoutPushdown, "SELECT id, CAST(c_decimal_25_2 AS TINYINT) from %s WHERE id = 1".formatted(table.getName())))
        .failure()
        .hasMessageContaining("Cannot cast '-129.49' to TINYINT");
assertThat(query("SELECT id, CAST(c_decimal_25_2 AS TINYINT) from %s WHERE id = 1".formatted(table.getName())))
        .failure()
        .hasMessageContaining("Bad value for type byte : -129");
```

https://github.com/trinodb/trino/pull/22951/files#diff-2f548c43d7eac10982d24f0ceca0ab7799cbadefe6df9d85af168cb606c7753aR322-R327